src/pybind/mgr/volumes/fs/nfs.py

```diff
@@ -0,0 +1,454 @@
+                'prefix': 'auth caps',
+                'entity': "client.%s" % (self.cluster_id),
+                'caps' : ['mon', 'allow *', 'osd', 'allow * pool=%s namespace=%s, allow rw tag cephfs data=a' % (self.pool_name, self.pool_ns), 'mds', 'allow * path=/'],
+                })
+
+            if ret!= 0:
+                return ret, out, err
+
+    def create_common_config(self, nodeid):
+        # TODO change rados url to "%url rados://{}/{}/{}".format(self.pool_name, self.pool_ns, nodeid)
+        result = """NFS_CORE_PARAM {{
+        Enable_NLM = false;
+        Enable_RQUOTA = false;
+        Protocols = 4;
+        }}
+
+        CACHEINODE {{
+        Dir_Chunk = 0;
+        NParts = 1;
+        Cache_Size = 1;
+        }}
+
+        NFSv4 {{
+        RecoveryBackend = rados_cluster;
+        Minor_Versions = 1, 2;
+        }}
+
+        RADOS_URLS {{
+        userid = {2};
+        }}
+
+        %url rados://{0}/{1}/export-1
+
+        RADOS_KV {{
+        pool = {0};
+        namespace = {1};
+        UserId = {2};
+        nodeid = {3};
+        }}""".format(self.pool_name, self.pool_ns, self.cluster_id, nodeid)
+        #self.ganeshaconf._write_raw_config(result, nodeid)
+
+        with self.mgr.rados.open_ioctx(self.pool_name) as ioctx:
+            if self.pool_ns:
+                ioctx.set_namespace(self.pool_ns)
+            ioctx.write_full(nodeid, result.encode('utf-8'))
+            log.debug(
+                    "write configuration into rados object %s/%s/%s:\n%s",
+                    self.pool_name, self.pool_ns, nodeid, result)
+
+    def create_instance(self):
+        assert self.ganeshaconf is not None
+        self.ganeshaconf = GaneshaConf(self)
+
+    def create_export(self):
+        assert self.ganeshaconf is not None
+        ex_id = self.ganeshaconf.create_export({
+            'path': "/",
+            'pseudo': "/cephfs",
+            'cluster_id': self.cluster_id,
+            'daemons': ["ganesha.a"],
+            'access_type': "RW",
+            'fsal': {"name": "CEPH", "user_id":self.cluster_id, "fs_name": "a", "sec_label_xattr": ""},
+            'clients': []
+            })
+        log.info("Export ID is {}".format(ex_id))
+        NFSConfig.exp_num += 1
+        #self.update_user_caps()
+        return 0, "", "Export Created Successfully"
+
+    def delete_export(self, ex_id):
+        assert self.ganeshaconf is not None
+        if not self.ganeshaconf.has_export(ex_id):
+            return 0, "No exports available",""
+        log.info("Export detected for id:{}".format(ex_id))
+        export = self.ganeshaconf.remove_export(ex_id)
+        self.ganeshaconf.reload_daemons(export.daemons)
+        return 0, "", "Export Deleted Successfully"
+
+    def check_fsal_valid(self):
+        fs_map = self.mgr.get('fs_map')
+        return [{'id': fs['id'], 'name': fs['mdsmap']['fs_name']}
+                for fs in fs_map['filesystems']]
+
+    def create_nfs_cluster(self, size):
```

Comment:
what does `size` mean in this context? It doesn't appear to be used in this function ... ?

Reply:
It denotes the number of daemons within a cluster. Only single cluster is created here for testing purpose. So I did not use it.